src/ring_theory/dedekind_domain.lean

```diff
@@ -0,0 +1,137 @@
+but are independent of that choice. The `..._iff` lemmas express this independence.
+
+## References
+
+* [D. Marcus, *Number Fields*][marcus1977number]
+* [J.W.S. Cassels, A. Frölich, *Algebraic Number Theory*][cassels1967algebraic]
+
+## Tags
+
+dedekind domain, dedekind ring
+-/
+
+variables (R A K : Type*) [comm_ring R] [integral_domain A] [field K]
+
+/-- A ring `R` has Krull dimension at most one if all nonzero prime ideals are maximal. -/
+def ring.dimension_le_one : Prop :=
+∀ p ≠ (⊥ : ideal R), p.is_prime → p.is_maximal
+
+open ideal ring
+
+lemma is_principal_ideal_ring.dimension_le_one
+  [is_principal_ideal_ring A] : ring.dimension_le_one A :=
+λ p nonzero prime, by { haveI := prime, exact is_prime.to_maximal_ideal nonzero }
+
+lemma integral_closure.dimension_le_one [nontrivial R] [algebra R A]
+  (h : dimension_le_one R) : dimension_le_one (integral_closure R A) :=
+begin
+  intros p ne_bot prime,
+  haveI := prime,
+  refine integral_closure.is_maximal_of_is_maximal_comap p
+    (h _ (integral_closure.comap_ne_bot ne_bot) _),
+  apply is_prime.comap
+end
+
+/--
+A Dedekind domain is a nonfield that is Noetherian, integrally closed, and
+has Krull dimension exactly one (`not_is_field` and `dimension_le_one`).
+
+The integral closure condition is independent of the choice of field of fractions:
+use `is_dedekind_domain_iff` to prove `is_dedekind_domain` for a given `fraction_map`.
+
+This is the default implementation, but there are equivalent definitions,
+`is_dedekind_domain_dvr` and `is_dedekind_domain_inv`.
+TODO: Prove that these are actually equivalent definitions.
+-/
+-- TODO: R should be a `comm_ring`
+class is_dedekind_domain : Prop :=
+(not_is_field : ¬ is_field A)
+(is_noetherian_ring : is_noetherian_ring A)
+(dimension_le_one : dimension_le_one A)
+(is_integrally_closed : integral_closure A (fraction_ring A) = ⊥)
+
+/-- The non-field, noetherian ring, dimension ≤ 1, integrally closed definition of Dedekind domain
+doesn't depend on the choice of fraction field. -/
+lemma is_dedekind_domain_iff (f : fraction_map A K) :
+  is_dedekind_domain A ↔
+    (¬ is_field A) ∧ is_noetherian_ring A ∧ dimension_le_one A ∧
+    integral_closure A f.codomain = ⊥ :=
+⟨λ ⟨hf, hr, hd, hi⟩, ⟨hf, hr, hd,
+  by rw [←integral_closure_map_alg_equiv (fraction_ring.alg_equiv_of_quotient f),
+         hi, algebra.map_bot]⟩,
+ λ ⟨hf, hr, hd, hi⟩, ⟨hf, hr, hd,
+  by rw [←integral_closure_map_alg_equiv (fraction_ring.alg_equiv_of_quotient f).symm,
+         hi, algebra.map_bot]⟩⟩
+
+/--
+A Dedekind domain is a nonfield that is Noetherian, and the localization at
+every nonzero prime is a discrete valuation ring.
+
+This is equivalent to `is_dedekind_domain`.
+TODO: prove the equivalence.
+-/
+structure is_dedekind_domain_dvr : Prop :=
+(not_is_field : ¬ is_field A)
+(is_noetherian_ring : is_noetherian_ring A)
+(is_dvr_at_nonzero_prime : ∀ P ≠ (⊥ : ideal A), P.is_prime →
+  discrete_valuation_ring (localization.at_prime P))
```

Comment:
I see. Is this a "temporary structure", i.e. one that will ultimately be replaced with a constructor taking those fields as input and spitting out a proof of `is_dedekind_domain A` and a lemma `integral_domain.is_dvr_at_nonzero_prime`? Is there a genuine need for this structure?

Reply:
I expect it will be temporary and replaced exactly as you described.

Reply:
I added replacing the structures as part of the dedekind domains project, but for now, it might be a bit easier to work with the structures.